File: .skills/openclaw-skills/skills/hinisal/qrcode-skills/scripts/decode.js
#!/usr/bin/env node
/**
 * QR Code Decoder - 本地 wechat-qr 解码 (Node.js 版)
 *
 * 用法:
 *   node scripts/decode.js <图片路径或URL>
 *   node scripts/decode.js --file <本地路径>
 *   node scripts/decode.js --url <图片URL>
 *
 * 输出 JSON:
 *   {"source": "wechat-qr", "contents": ["..."]}
 *   {"error": "..."}
 */

const https = require("https");
const http = require("http");
const fs = require("fs");
const os = require("os");
const path = require("path");

function output(source, contents) {
  console.log(JSON.stringify({ source, contents }));
  process.exit(0);
}

function error(msg) {
  console.log(JSON.stringify({ error: msg }));
  process.exit(1);
}

function isUrl(s) {
  return s.startsWith("http://") || s.startsWith("https://");
}

function downloadToTemp(url) {
  return new Promise((resolve, reject) => {
    const ext = path.extname(url.split("?")[0]) || ".png";
    const tmp = path.join(os.tmpdir(), `qr_${Date.now()}${ext}`);
    const mod = url.startsWith("https") ? https : http;
    const file = fs.createWriteStream(tmp);
    mod.get(url, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        file.close();
        fs.unlinkSync(tmp);
        return downloadToTemp(res.headers.location).then(resolve, reject);
      }
      res.pipe(file);
      file.on("finish", () => { file.close(); resolve(tmp); });
    }).on("error", (e) => { file.close(); if (fs.existsSync(tmp)) fs.unlinkSync(tmp); reject(e); });
  });
}

async function decodeWithWechat(imagePath) {
  try {
    const { scan } = await import("qr-scanner-wechat");
    const sharp = require("sharp");
    const { data, info } = await sharp(imagePath)
      .ensureAlpha()
      .raw()
      .toBuffer({ resolveWithObject: true });
    const result = await scan({
      data: Uint8ClampedArray.from(data),
      width: info.width,
      height: info.height,
    });
    return result?.text ? [result.text] : null;
  } catch {
    return null;
  }
}

async function main() {
  const argv = process.argv.slice(2);

  if (!argv.length) error("用法: node decode.js <图片路径或URL>");

  let mode, target;
  if ((argv[0] === "--file" || argv[0] === "--url") && argv.length >= 2) {
    mode = argv[0];
    target = argv[1];
  } else {
    target = argv[0];
    mode = isUrl(target) ? "--url" : "--file";
  }

  if (mode === "--file") {
    if (!fs.existsSync(target)) error(`文件不存在: ${target}`);

    const results = await decodeWithWechat(target);
    if (results) output("wechat-qr", results);

    error("无法解码: 本地 wechat-qr 未识别到二维码");
  } else {
    let tmpPath = null;
    try {
      tmpPath = await downloadToTemp(target);
      const results = await decodeWithWechat(tmpPath);
      if (results) output("wechat-qr", results);
    } catch (e) {
      error(`下载图片失败: ${e.message}`);
    } finally {
      if (tmpPath && fs.existsSync(tmpPath)) fs.unlinkSync(tmpPath);
    }

    error("无法解码: 本地 wechat-qr 未识别到二维码");
  }
}

main();

File: .skills/openclaw-skills/skills/hinisal/qrcode-skills/scripts/decode.py
"""
QR Code Decoder - 本地 zxing 解码。

用法:
  python scripts/decode.py <图片路径或URL>
  python scripts/decode.py --file <本地文件路径>
  python scripts/decode.py --url <图片URL>

输出 JSON:
  {"source": "zxing", "contents": ["..."]}
  错误时: {"error": "..."}
"""

import sys
import json
import os
import tempfile
from pathlib import Path


def is_url(s: str) -> bool:
    return s.startswith("http://") or s.startswith("https://")


def download_image(url: str) -> str:
    """下载图片到临时文件，返回临时文件路径。"""
    import urllib.request

    suffix = Path(url.split("?")[0]).suffix or ".png"
    tmp = tempfile.NamedTemporaryFile(delete=False, suffix=suffix)
    try:
        urllib.request.urlretrieve(url, tmp.name)
    except Exception as e:
        tmp.close()
        os.unlink(tmp.name)
        raise RuntimeError(f"下载图片失败: {e}")
    tmp.close()
    return tmp.name


def decode_with_zxing(image_path: str) -> list[str] | None:
    """使用 zxingcpp 本地解码，成功返回内容列表，失败返回 None。"""
    try:
        import zxingcpp
        from PIL import Image

        img = Image.open(image_path)
        results = zxingcpp.read_barcodes(img)
        if results:
            return [r.text for r in results]
        return None
    except ImportError:
        return None
    except Exception:
        return None


def output(source: str, contents: list[str]):
    print(json.dumps({"source": source, "contents": contents}, ensure_ascii=False))
    sys.exit(0)


def error(msg: str):
    print(json.dumps({"error": msg}, ensure_ascii=False))
    sys.exit(1)


def main():
    if len(sys.argv) < 2:
        error("用法: python decode.py <图片路径或URL>")

    args = sys.argv[1:]

    if not args:
        error("用法: python decode.py <图片路径或URL>")

    arg1 = args[0]
    if arg1 in ("--file", "--url") and len(args) >= 2:
        mode = arg1
        target = args[1]
    else:
        target = arg1
        mode = "--url" if is_url(target) else "--file"

    if mode == "--file":
        if not os.path.isfile(target):
            error(f"文件不存在: {target}")

        results = decode_with_zxing(target)
        if results:
            output("zxing", results)

        error("无法解码: 本地 zxing 未识别到二维码")

    else:  # --url
        tmp_path = None
        try:
            tmp_path = download_image(target)
            results = decode_with_zxing(tmp_path)
            if results:
                output("zxing", results)
        except RuntimeError as e:
            error(str(e))
        finally:
            if tmp_path and os.path.exists(tmp_path):
                os.unlink(tmp_path)

        error("无法解码: 本地 zxing 未识别到二维码")


if __name__ == "__main__":
    main()
